rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll

```diff
@@ -183,7 +179,26 @@ class CastExprTree extends StandardPostOrderTree instanceof CastExpr {
   override AstNode getChildNode(int i) { i = 0 and result = super.getExpr() }
 }
 
-class ClosureExprTree extends LeafTree instanceof ClosureExpr { }
+// Closures have their own CFG scope, so we need to make sure that their
+// CFG is not mixed with the surrounding CFG. This is done by retrofitting
+// `first`, `propagatesAbnormal`, and `succ` below.
+class ClosureExprTree extends StandardPostOrderTree, ClosureExpr {
```

Comment:
I'm not sure I like the way this is formulated. If I understand correctly you want the `succ` relation of a `StandardOrderTree` and the `first` and `last` predicates of  a `LeafTree`. Perhaps it would be clearer to make this class extend `StandardOrderTree` and avoid the `not succ = this` restriction in `succ` and add an explicit `last` predicate. It's about the same amount of lines and you avoid the tweaks to "undo" the behaviour of a `Post` order tree.